File: src/types/workflow.ts
import type { AgentNodeData } from "./agent";
import type { Node, Edge } from "@xyflow/react";

export type AgentNode = Node<AgentNodeData, "agent">;

export interface WorkflowConnection {
  id: string;
  sourceAgentId: string;
  targetAgentId: string;
  label?: string;
  edgeKind?: "dataflow" | "memory" | "feedback" | "control";
}

export interface ExecutionSettings {
  maxParallel: number;
  timeoutSeconds: number;
  retryOnFailure: boolean;
  maxRetries: number;
}

export interface WorkflowMeta {
  name: string;
  version: string;
  description: string;
  projectRoot: string;
  createdAt: string;
  updatedAt: string;
}

export interface WorkflowDef {
  meta: WorkflowMeta;
  agents: AgentNodeData[];
  connections: WorkflowConnection[];
  executionSettings: ExecutionSettings;
  nodePositions: Record<string, { x: number; y: number }>;
}

export interface WorkflowState {
  nodes: AgentNode[];
  edges: Edge[];
  meta: WorkflowMeta;
  executionSettings: ExecutionSettings;
  isDirty: boolean;
  filePath: string | null;
}

/** Validation result for a workflow graph. */
export interface ValidationResult {
  valid: boolean;
  errors: ValidationError[];
  warnings: ValidationWarning[];
}

export interface ValidationError {
  nodeId?: string;
  message: string;
  kind: "cycle" | "disconnected" | "missing_prompt" | "no_model";
}

export interface ValidationWarning {
  nodeId?: string;
  message: string;
  kind: "high_token_budget" | "many_tools" | "no_hooks_on_bash";
}

File: src-tauri/src/models/workflow.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct WorkflowMeta {
    pub name: String,
    pub version: String,
    pub description: String,
    #[serde(rename = "projectRoot")]
    pub project_root: String,
    #[serde(rename = "createdAt")]
    pub created_at: String,
    #[serde(rename = "updatedAt")]
    pub updated_at: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ExecutionSettings {
    #[serde(rename = "maxParallel")]
    pub max_parallel: u32,
    #[serde(rename = "timeoutSeconds")]
    pub timeout_seconds: u32,
    #[serde(rename = "retryOnFailure")]
    pub retry_on_failure: bool,
    #[serde(rename = "maxRetries")]
    pub max_retries: u32,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct NodePosition {
    pub x: f64,
    pub y: f64,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct WorkflowDef {
    pub meta: WorkflowMeta,
    pub agents: Vec<serde_json::Value>,
    pub connections: Vec<serde_json::Value>,
    #[serde(rename = "executionSettings")]
    pub execution_settings: ExecutionSettings,
    #[serde(rename = "nodePositions")]
    pub node_positions: HashMap<String, NodePosition>,
}
